CEL expressions approver design (#256)
cert-manager/approver-policy: design/20230726-cel-policy.md

- [ ] This design doc has been discussed and approved
- [ ] Test plan has been agreed upon and the tests implemented 
- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]

## Summary

<!--
This section is important for producing high-quality, user-focused
documentation such as release notes.

A good summary is probably around a paragraph in length.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

While approver-policy allows policies to include namespace selectors, this does not scale when attempting to
enforce CSR attributes that must be a function of the CertificateRequest namespace (or other CertificateRequest fields).
This proposal suggests an extension to the CertificateRequestPolicy API allowing policy authors to specify approval rules using

> [CEL expressions](https://github.com/google/cel-spec) containing variables with information from the CertificateRequest.

Reviewer: Is there a reason to point to that instead of https://kubernetes.io/docs/reference/using-api/cel/?

Author: We could link to both. The Kubernetes docs give a good overview, but if you're interested in the details, I think you need to consult the detailed spec.

Author: Good point, thanks! I will change the link in my next round on this. Also thanks for the rewording suggestions! 👍

- validate CEL expressions in CertificateRequestPolicy on admission
- extensions to the CertificateRequestPolicy API should be backwards compatible

### Non-Goals

<!--
What is out of scope for this proposal? Listing non-goals helps to focus discussion
and make progress.
-->

- CEL support in CertificateRequestPolicy `constraints`

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
What is the desired outcome and how do we measure success?
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation- those should go into "Design Details" below.
-->

The current CertificateRequestPolicy API allows a policy author to specify `allowed` CSR attributes in the
CertificateRequest, and will deny the request if it contains an attribute which is not present in the `allowed` block.
To configure valid CSR attribute values, the API provides fields for allowed
`values` (array; any element matching the attribute is sufficient),
`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the request attribute.

> Allowed string fields accept wildcards "*" within its value(s).

Reviewer: ```suggestion
Allowed string fields accept wildcards `"*"` within its value(s).
```
? Or `*` ...

This proposal suggests to add support for specifying allowed string fields using CEL expressions, similar to the existing
wildcard support. While CEL expressions could eventually replace the current CSR attribute validation API - including the 
wildcard support, the main improvement would be support for policies that specifies an allowed attribute value that depends
on CertificateRequest fields e.g. `namespace`. If implemented, this new feature would close
[#62](https://github.com/cert-manager/approver-policy/issues/62).

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this proposal gets implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### X.509 DNS SAN in namespace domain

As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
self-provisioning of server certificates. Since the cluster is multi-tenant, I want to ensure that tenants do not

> "tip on each other's toes" and enforce that the issuer only signs certificates where the X.509 DNS SAN attribute

Reviewer: ```suggestion
"trip on each other's toes" and enforce that the issuer only signs certificates where the X.509 DNS SAN attribute
```

ends with `<namespace>.<domain>`.

A variant of this user story is exemplified in [#62](https://github.com/cert-manager/approver-policy/issues/62),
and I think it's one of the most common user stories the proposed feature will solve.

#### SPIFFE ID identifying namespace

As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing

> self-provisioning of X.509 SVID certificates. The SPIFFE ID (X.509 URI SAN) should identify the namespace

Reviewer: It'd be nice if `SVID` and `SPIFFE ID` were links to things.

Author: Thanks @jsoref! Updated with links.

today, but with the following extension: If any value **OR** expression matches the corresponding CSR attribute,
the attribute will be allowed.

For the CEL expressions, we need to decide if they should either:

1. Emit a string that will be used as values do. Example: `'*.%s.svc'.format(cr.namespace)`
2. Return a bool indicating if the expression allows the value. Example: `self.endsWith(cr.namespace + '.svc')` (preferred)

To use expressions in CertificateRequestPolicy the API may look like this (example):

```yaml
spec:
  allowed:
    dnsNames:
      expressions:
        - "self.endsWith(cr.namespace + '.svc')"
        - "self.endsWith(cr.namespace + '.svc.cluster.local')"
```

The CEL context variable `self` is inspired by
[CEL expressions in Kubernetes](https://kubernetes.io/docs/reference/using-api/cel/),
and should appear familiar to our users. CEL expressions returning a bool will resemble Kubernetes CEL validation
expressions and is preferred for this reason (IMO). 

It should be allowed to mix `values` and `expressions`, so the following policy should also be acceptable:

```yaml
spec:
  allowed:
    dnsNames:

>       expressions:
>         - "self.endsWith(cr.namespace + '.apps.my-cluster.com')"

Author: ```suggestion
      validations:
        - rule: "self.endsWith(cr.namespace + '.apps.my-cluster.com')"
          message: "DNSName should be in cluster apps domain"
```

After a bit of thinking, I think we should align our design closer to upstream [validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules).

I'll update the suggested design now.

Author: Modified the proposal to reuse the Kubernetes CEL validations.

The [SPIFFE ID](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md) (X.509 URI SAN) should identify
the namespace (and service account), and must start with `spiffe://<trust-domain>/ns/<namespace>/sa/`.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go into as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes/PKI ecosystem.

-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

An [experimental approver-policy plugin](https://github.com/erikgb/cel-approver-policy-plugin) providing some of the

> requested features exists, and it's proposed to base the design on this plugin.

Reviewer: this line is awkward. https://en.wikipedia.org/wiki/Garden-path_sentence

In general, I'd discourage the use of abbreviations (`'`), 
```suggestion
requested features exists, and we propose to base the design on this plugin.
```

information to express the idea and why it was not acceptable.
-->

### CEL expressions with full decoded CSR variable

This alternative was suggested by @inteon on Slack.

The idea here is to provide the policy author with a CEL context variable containing the decoded CSR replacing the
`self` variable in the proposed design. Since the expressions now get access to the full CSR, the expressions will have
to be "promoted" in the CertificateRequestPolicy API. The expression will have to return a bool indicating approved or
denied. An example policy with this alternative API might look like (example);

```yaml
spec:
  expressions:
    - >-
      has(csr.dnsNames) &&
      csr.dnsNames.all(dnsName,
      dnsName.endsWith(cr.namespace + '.svc') ||
      dnsName.endsWith(cr.namespace + '.svc.cluster.local'))
```

Pros:

- Extremely flexible. I.e. allows for CSR cross-attribute validation.
- Some would find this API more expressive and better describe the intentions.

Cons:

- More complex expressions that will require more in-depth knowledge of CEL.
- Will create a "split" API, since you would probably use either `expressions` or `allowed`/`constraints`.

>   This could increase the maintenance burden and source to subtle bugs. Breaking the API could be an option to mend this.

Reviewer: ```suggestion
  This could increase the maintenance burden and lead to subtle bugs. Breaking the API could be an option to mitigate this.
```

- Will probably require custom CEL functions to fulfill the requirements in approver-policy. This is the main reason for
  recommending this design, and will be elaborated in the following.

Let me start with a quote from the
[approver-policy documentation](https://cert-manager.io/docs/projects/approver-policy/#allowed):

> Allowed is the block that defines attributes that match against the corresponding attribute in the request.
> A request is permitted by the policy if the request omits an allowed attribute, but will deny the request if
> it contains an attribute which is not present in the allowed block.

The last sentence describes the requirement that could need custom CEL function(s) to allow the user to specify allowed
CSR attributes. We cannot expect the user to include validations for **all** CSR attributes in the expression(s). So I

> think we at least need a function to list allowed CSR attribute usage.

Author: I have now looked at the source code, and approver-policy only evaluates a policy against the CSR attributes that approver-policy supports. So you are correct that it might be considered too permissive in this sense. I suggest creating an issue for this if it is something that we should attempt to fix. Changing it would be breaking tho. 😉

<!--
This template is adapted from Kubernetes Enchancements KEP template https://raw.githubusercontent.com/kubernetes/enhancements/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md
-->

> # CEL validation rules to approve or deny CertificateRequest

Reviewer: ```suggestion
# CEL validation rules to approve or deny a CertificateRequest
```

  - [CEL expressions with full decoded CSR variable](#cel-expressions-with-full-decoded-csr-variable)
<!-- /toc -->

## Release Signoff Checklist

This checklist contains actions which must be completed before a PR implementing this design can be merged.

- [ ] This design doc has been discussed and approved
- [ ] Test plan has been agreed upon and the tests implemented 
- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]

## Summary

<!--
This section is important for producing high-quality, user-focused
documentation such as release notes.

A good summary is probably around a paragraph in length.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

While approver-policy allows policies to include namespace selectors, this does not scale when attempting to

> enforce CSR attributes that must be a function of the CertificateRequest namespace (or other CertificateRequest fields).

Reviewer: ```suggestion
enforce CSR attributes that must be a function of the CertificateRequest's namespace (or other CertificateRequest fields).
```

This proposal suggests an extension to the CertificateRequestPolicy API allowing policy authors to specify approval rules using
[CEL in Kubernetes](https://kubernetes.io/docs/reference/using-api/cel/) using variables with information from the CertificateRequest.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
the proposed enhancement.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to
demonstrate the interest in this functionality amongst the community.
-->

### Goals

<!--
List specific goals. What is this proposal trying to achieve? How will we
know that this has succeeded?
-->

> - allow policy authors to express `allowed` CSR attribute value(s) to be a function of CertificateRequest fields using CEL expressions

Reviewer: ```suggestion
- allow policy authors to express `allowed` CSR attribute value(s) to be a function of CertificateRequest's fields using CEL expressions
```

- include variables with information from the CertificateRequest and CSR attribute(s) in the CEL context provided to expressions
- validate CEL expressions in CertificateRequestPolicy on admission
- extensions to the CertificateRequestPolicy API should be backwards compatible

### Non-Goals

<!--
What is out of scope for this proposal? Listing non-goals helps to focus discussion
and make progress.
-->

- CEL support in CertificateRequestPolicy `constraints`

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
What is the desired outcome and how do we measure success?
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation- those should go into "Design Details" below.
-->

The current CertificateRequestPolicy API allows a policy author to specify `allowed` CSR attributes in the
CertificateRequest, and will deny the request if it contains an attribute which is not present in the `allowed` block.
To configure valid CSR attribute values, the API provides fields for allowed
`values` (array; any element matching the attribute is sufficient),

> `value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the request attribute.

Reviewer: possibly?
```suggestion
`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the requested attribute.
```

Allowed string fields accept wildcards `"*"` within its value(s).

This proposal suggests to add support for specifying allowed string fields using CEL expressions, similar to the existing
wildcard support. While CEL expressions could eventually replace the current CSR attribute validation API - including the 

> wildcard support, the main improvement would be support for policies that specifies an allowed attribute value that depends

Reviewer: ```suggestion
wildcard support, the main improvement would be support for policies that specify an allowed attribute value that depends
```

on CertificateRequest fields, e.g. `namespace`. If implemented, this new feature would close
[#62](https://github.com/cert-manager/approver-policy/issues/62).

An [experimental approver-policy plugin](https://github.com/erikgb/cel-approver-policy-plugin) providing some of the
requested features exists, and the proposal (including alternatives) is inspired by this plugin.

> The main downside of using a plugin for this, in addition increased maintenance/complexity,

Reviewer: ```suggestion
The main downside of using a plugin for this, in addition to increased maintenance/complexity,
```

is that the approver-policy rules are still in play.

> Which means that approver rules are somehow duplicated in the `allowed` block and in the plugin configuration and
> should be avoided.

Reviewer: ```suggestion
Which means that approver rules are somehow duplicated in the `allowed` block and in the plugin configuration but
this should be avoided as maintaining things in multiple places is error prone.
```

change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

The proposal is to add sibling properties to `values` and `value` properties inside the `allowed` block. The new
fields should contain CEL validations used to validate CSR attributes with similar
[behavior](https://cert-manager.io/docs/projects/approver-policy/#allowed) as approver-policy has
today, but with the following extension: If validations are present for the given CSR attribute, and **ALL**
validations passes, the attribute will be allowed. This also holds if no values are listed.

The validations themselves should be expressed using the same API as Kubernetes `x-kubernetes-validations`, having
a mandatory `rule` field for the CEL expression and an optional `message` field for the validation error message.

To use validations in CertificateRequestPolicy the API may look like this (example):

```yaml
spec:
  allowed:
    dnsNames:
      validations:
        - message: only dnsNames in the local namespace service domain is allowed.

>           rule: self.endsWith(cr.namespace + '.svc') || self.endsWith(cr.namespace + '.svc.cluster.local')

Reviewer 2: I would start with a `cr` value that only contains the basic fields of the CertifiicateRequest resource (name, namespace, annotations, labels, uuid, ....). Once that works, we could also add other properties (like x509 stuff).

Author: Agreed, I think we should start even simpler in the initial implementation and provide just CR `namespace` and `name`. I will add this initial MVP to the proposed design.

Author: Updated, PTAL!

```

The CEL context variable `self` is inspired by
[CEL expressions in Kubernetes](https://kubernetes.io/docs/reference/using-api/cel/),
and should appear familiar to our users. 

It should be allowed to mix `values` and `validations`, so the following policy should also be acceptable:

```yaml
spec:
  allowed:
    dnsNames:
      validations:
        - message: dnsName can be maximum 64 characters.

>           rule: self.size() =< 64

Reviewer 2: How could we limit the number of dnsNames? Could we eg. have a "selfs" parameter?

Author: This is an interesting and important comment! Strictly speaking, many CSR attributes are multi-valued - so it could make sense to make `self` refer to the list of attribute values in the CSR - which will allow for the validation you are requesting here. But I suspect this will make the feature less user-friendly. The validation example this review comment refers to would then have to be expressed as `self.all(dnsName, dnsName.size() =< 64)` - which is not too bad IMO.

But this could easily lead to user errors like `self.size() =< 64` would now limit the number of dnsNames in the CSR to 63. 😉 Which is probably not what the user had in mind.

Author: IMO the most user-friendly is to keep the validations under `allowed`  as per-item validations. This is how the `value`/`values` fields in the current API work. More complex CEL validations, typically cross-cutting concerns could be introduced later outside of the `allowed` API structure. I am updating the proposal with this clarification now.

-->

### Will enabling / using this feature result in new API calls (i.e to Kubernetes apiserver or external services)?
<!--
We should ensure that cert-manager does not hammer any external services with excessive calls.
Consider whether there will be sufficient backoff if any external calls fail and need to be retried.
-->

### Will enabling / using this feature result in increasing size or count of the existing API objects?

<!--
For example, will cert-manager `CustomResourceDefinition`s increase in size, will there be more `Secret`s or `CertificateRequest`s created?
-->

### Will enabling / using this feature result in significant increase of resource usage? (CPU, RAM...)

<!--
For example, will implementing this feature result in more objects being cache thus increasing memory consumption?
-->

## Drawbacks

<!--
Why should this proposal _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

> ### CEL expressions with similar semantics as values

Reviewer: ```suggestion
### CEL expressions with semantics similar to values
```

> _Note: This was the original proposal, after some consideration it is considered inferior to the current proposal,

Reviewer: ```suggestion
_Note: This was the original proposal, after some consideration it was considered inferior to the current proposal,
```

mainly because we think we should align this CEL API extension more towards how Kubernetes supports CEL._

The proposal is to add sibling properties to `values` and `value` properties inside the `allowed` block. The new
fields should contain CEL expressions used to validate CSR attributes with the same
[behavior](https://cert-manager.io/docs/projects/approver-policy/#allowed) as approver-policy has
today, but with the following extension: If any value **OR** expression matches the corresponding CSR attribute,
the attribute will be allowed.

For the CEL expressions, we need to decide if they should either:

1. Emit a string that will be used as values do. Example: `'*.%s.svc'.format(cr.namespace)`
2. Return a bool indicating if the expression allows the value. Example: `self.endsWith(cr.namespace + '.svc')` (preferred)

To use expressions in CertificateRequestPolicy the API may look like this (example):

```yaml
spec:
  allowed:
    dnsNames:
      expressions:
        - "self.endsWith(cr.namespace + '.svc')"

>         - "self.endsWith(cr.namespace + '.svc.cluster.local')"

Reviewer 2: Maybe, it would be useful if we could also express that eg. you can have at most 3 dnsNames in your cert?

```

The CEL context variable `self` is inspired by
[CEL expressions in Kubernetes](https://kubernetes.io/docs/reference/using-api/cel/),
and should appear familiar to our users. CEL expressions returning a bool will resemble Kubernetes CEL validation
expressions and is preferred for this reason (IMO).

It should be allowed to mix `values` and `expressions`, so the following policy should also be acceptable:

```yaml
spec:
  allowed:
    dnsNames:
      expressions:
        - "self.endsWith(cr.namespace + '.apps.my-cluster.com')"
      values:
        - "*.sub.domain.com"
```

### CEL expressions with full decoded CSR variable

This alternative was suggested by @inteon on Slack.

The idea here is to provide the policy author with a CEL context variable containing the decoded CSR replacing the
`self` variable in the proposed design. Since the expressions now get access to the full CSR, the expressions will have
to be "promoted" in the CertificateRequestPolicy API. The expression will have to return a bool indicating approved or

> denied. An example policy with this alternative API might look like (example);

Reviewer: ```suggestion
denied. An example policy with this alternative API might look like (example):
```

- extensions to the CertificateRequestPolicy API should be backwards compatible

### Non-Goals

<!--
What is out of scope for this proposal? Listing non-goals helps to focus discussion
and make progress.
-->

- CEL support in CertificateRequestPolicy `constraints`
- CEL variable with decoded CSR in the CEL context provided to expressions

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
What is the desired outcome and how do we measure success?
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation- those should go into "Design Details" below.
-->

The current CertificateRequestPolicy API allows a policy author to specify `allowed` CSR attributes in the
CertificateRequest, and will deny the request if it contains an attribute which is not present in the `allowed` block.
To configure valid CSR attribute values, the API provides fields for allowed
`values` (array; any element matching the attribute is sufficient),
`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the requested attribute.

> Allowed string fields accept wildcards `"*"` within its value(s).

Reviewer: ```suggestion
Allowed string fields accept wildcards `"*"` within their values.
```
I think...

> This proposal suggests to add support for specifying allowed string fields using CEL expressions, similar to the existing

Reviewer: ```suggestion
This proposal suggests adding support for specifying allowed string fields using CEL expressions, similar to the existing
```

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go into as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

N/A

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes/PKI ecosystem.

-->

N/A

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

The proposal is to add sibling properties to `values` and `value` properties inside the `allowed` block.
The new fields should contain CEL validations used to validate CSR attribute values with similar
[behavior](https://cert-manager.io/docs/projects/approver-policy/#allowed) as approver-policy has today,
but with the following extension: If validations are present for the given CSR attribute, and **ALL**

> validations passes for **ALL* attribute values, the attribute will be allowed.

Reviewer: ```suggestion
validations pass for **ALL* attribute values, the attribute will be allowed.
```

Note: To ensure we don't break behaviors in the existing API version, the following should be enforced:

> * if `required` set to `true` in policy and attribute value empty in request, the request will be denied
>   by policy without running any validations.
> * if `required` set to `true` in policy, the policy will need specify `values`/`value` or `validations`.

Reviewer: ```suggestion
* if `required` is set to `true` in policy and attribute value empty in request, the request will be denied
  by policy without running any validations.
* if `required` is set to `true` in policy, the policy will need specify `values`/`value` or `validations`.
```
